### PR TITLE
Remove QtConcurrent feature

### PR DESCRIPTION
**Pull request by andreadotti**
The use of `QtConcurrent::blockingMap` is replaced
with `tbb::parallel_for`.
To note: the setting of the number of threads in tbb
is deprecated. Thus the code now ignores the setting
of this parameter.
The function calls specifying the number of threads
are kept for back-compatibility but the parameter
is ignored.

### CODE REVIEW

**Comment by dgirardeau** [comment]
Ah sorry, big misunderstanding: I thought TBB would remain optional! Like, we would have the option in CMake to use Qt or TBB. And not simply replace Qt by TBB...

**Comment by dgirardeau** [comment]
Yes I was expecting this kind of solution actually ;)

File: include/DgmOctree.h [hunk 1]
```diff
@@ -21,11 +21,11 @@
 //#define TEST_CELLS_FOR_SPHERICAL_NN
 
 #ifndef CC_DEBUG
-#ifdef CC_CORE_LIB_USES_QT_CONCURRENT
```
Comment (dgirardeau) [inline include/DgmOctree.h:24]: Shouldn't your restore that as well? (I don't want to compile the official Windows version with TBB 😉  )

**Review by dgirardeau — CHANGES_REQUESTED** [review]

File: include/DgmOctree.h (hunk 1 above)
Comment (andreadotti) [reply include/DgmOctree.h:24]: Sorry, I forgot to convert the PR to draft before starting to work on this. Should be correct now

File: include/DgmOctree.h [hunk 2]
```diff
@@ -1090,23 +1090,23 @@ namespace CCCoreLib
...
+			\param maxThreadCount the maximum number of threads to use (0 = all). Ignored if 'multiThread' is false or with tbb. 
 			\return the number of processed cells (or 0 is something went wrong)
 		**/
 		unsigned executeFunctionForAllCellsAtLevel(	unsigned char level,
 													octreeCellFunc func,
 													void** additionalParameters,
 													bool multiThread = false,
 													GenericProgressCallback* progressCb = nullptr,
-													const char* functionTitle = nullptr,
+													const char* functionTitle = nullptr,
```
Comment (dgirardeau) [inline include/DgmOctree.h:1109]: is this a spurious space character at the end?

File: src/DgmOctree.cpp [hunk 3]
```diff
@@ -20,10 +20,17 @@
...
 #ifdef ENABLE_MT_OCTREE
+#if defined(CC_CORE_LIB_USES_QT_CONCURRENT)
 #include <QThread>
 #include <QThreadPool>
 #include <QtConcurrentMap>
+#elif defined(CC_CORE_LIB_USES_TBB)
+#include <algorithm>
+#include <tbb/parallel_for.h>
+#else
+#error "Multithreaded OCtree should be enabled only with Qt OR TBB!"
```
Comment (dgirardeau) [inline src/DgmOctree.cpp:31]: Octree (without capital C)

File: src/DgmOctree.cpp [hunk 4]
```diff
@@ -3300,14 +3307,14 @@ unsigned DgmOctree::executeFunctionForAllCellsAtLevel(	unsigned char level,
 														bool multiThread/*=false*/,
 														GenericProgressCallback* progressCb/*=0*/,
 														const char* functionTitle/*=0*/,
-														int maxThreadCount/*=0*/)
+														int maxThreadCount/*=0*/)
```
Comment (dgirardeau) [inline src/DgmOctree.cpp:3310]: spurious char at the end?

File: src/DgmOctree.cpp [hunk 5]
```diff
@@ -3300,14 +3307,14 @@ unsigned DgmOctree::executeFunctionForAllCellsAtLevel(	unsigned char level,
...
-														int maxThreadCount/*=0*/)
+														int maxThreadCount/*=0*/) 
 {
 	if (m_thePointsAndTheirCellCodes.empty())
 		return 0;
 
 #ifdef ENABLE_MT_OCTREE
 
-	//cells that will be processed by QtConcurrent::map
+	//cells that will be processed by tbb::parallel_for
```
Comment (dgirardeau) [inline src/DgmOctree.cpp:3317]: QtConcurrent::map or tbb::parallel_for

File: src/DgmOctree.cpp [hunk 6]
```diff
@@ -3502,14 +3509,20 @@ unsigned DgmOctree::executeFunctionForAllCellsAtLevel(	unsigned char level,
 		s_jumps = 0.0;
 		s_binarySearchCount = 0.0;
 #endif
-
+#ifdef CC_CORE_LIB_USES_QT_CONCURRENT
+		// The QtConcurrent is used in case both Qt and TBB are available
```
Comment (dgirardeau) [inline src/DgmOctree.cpp:3513]: Isn't it only if Qt is available?

(and I would remove 'The' in front of QtConcurrent)

File: src/DgmOctree.cpp [hunk 7]
```diff
@@ -3502,14 +3509,20 @@ unsigned DgmOctree::executeFunctionForAllCellsAtLevel(	unsigned char level,
...
 		if (maxThreadCount == 0)
 		{
 			maxThreadCount = QThread::idealThreadCount();
 		}
 		QThreadPool::globalInstance()->setMaxThreadCount(maxThreadCount);
 		QtConcurrent::blockingMap(cells, [this](const octreeCellDesc& desc) { m_MT_wrapper.launchOctreeCellFunc(desc); } );
-
+#else //It uses TBB
```
Comment (dgirardeau) [inline src/DgmOctree.cpp:3520]: //Using TBB

File: src/DgmOctree.cpp [hunk 8]
```diff
@@ -3568,7 +3581,7 @@ unsigned DgmOctree::executeFunctionForAllCellsStartingAtLevel(unsigned char star
 
 #ifdef ENABLE_MT_OCTREE
 
-	//cells that will be processed by QtConcurrent::map
+	//cells that will be processed by tbb::parallel_for
```
Comment (dgirardeau) [inline src/DgmOctree.cpp:3584]: QtConcurrent::map or tbb::parallel_for

File: src/DgmOctree.cpp [hunk 9]
```diff
@@ -4052,14 +4065,20 @@ unsigned DgmOctree::executeFunctionForAllCellsStartingAtLevel(unsigned char star
 		s_jumps = 0.0;
 		s_binarySearchCount = 0.0;
 #endif
-
+#ifdef CC_CORE_LIB_USES_QT_CONCURRENT
+		// QtConcurrent in case both Qt and TBB are available
```
Comment (dgirardeau) [inline src/DgmOctree.cpp:4069]: Same as above: isn't it only with Qt?

File: src/DgmOctree.cpp [hunk 10]
```diff
@@ -4052,14 +4065,20 @@ unsigned DgmOctree::executeFunctionForAllCellsStartingAtLevel(unsigned char star
...
 		if (maxThreadCount == 0)
 		{
 			maxThreadCount = QThread::idealThreadCount();
 		}
 		QThreadPool::globalInstance()->setMaxThreadCount(maxThreadCount);
 		QtConcurrent::blockingMap(cells, [this](const octreeCellDesc& desc) { m_MT_wrapper.launchOctreeCellFunc(desc); } );
-
+#else
```
Comment (dgirardeau) [inline src/DgmOctree.cpp:4076]: maybe 'elseif #CC_CORE_LIB_USES_TBB' for completeness

File: src/DistanceComputationTools.cpp [hunk 11]
```diff
@@ -20,15 +20,20 @@
 #include <algorithm>
 #include <cassert>
 
-#ifdef CC_CORE_LIB_USES_QT_CONCURRENT
 #ifndef CC_DEBUG
-//enables multi-threading handling
```
Comment (dgirardeau) [inline src/DistanceComputationTools.cpp:25]: maybe restore this comment? (just before #define ENABLE_CLOUD2MESH_DIST_MT)

Comment (dgirardeau) [reply src/DistanceComputationTools.cpp:25]: //enables multi-threading handling with Qt

File: src/DistanceComputationTools.cpp [hunk 12]
```diff
@@ -20,15 +20,20 @@
...
+#if defined(CC_CORE_LIB_USES_QT_CONCURRENT)
 #define ENABLE_CLOUD2MESH_DIST_MT
-
 #include <QtConcurrentMap>
 #include <QtCore>
+#elif defined(CC_CORE_LIB_USES_TBB)
+//enables multi-threading handling
```
Comment (dgirardeau) [inline src/DistanceComputationTools.cpp:29]: //enables multi-threading handling with TBB

File: src/DistanceComputationTools.cpp [hunk 13]
```diff
@@ -1126,7 +1131,11 @@ static DistanceComputationTools::Cloud2MeshDistanceComputationParams s_params_MT
 //'processTriangles' mechanism (based on bit mask)
 static std::vector<std::vector<bool>*> s_bitArrayPool_MT;
 static bool s_useBitArrays_MT = true;
+#ifdef CC_CORE_LIB_USES_QT_CONCURRENT
 static QMutex s_currentBitMaskMutex;
+#else
```
Comment (dgirardeau) [inline src/DistanceComputationTools.cpp:1136]: #elif defined(CC_CORE_LIB_USES_TBB)?

File: src/DistanceComputationTools.cpp [hunk 14]
```diff
@@ -1778,13 +1787,21 @@ int DistanceComputationTools::computeCloud2MeshDistanceWithOctree(	OctreeAndMesh
...
 
+#ifdef CC_CORE_LIB_USES_QT_CONCURRENT
 		int maxThreadCount = params.maxThreadCount;
 		if (maxThreadCount == 0)
 		{
 			maxThreadCount = QThread::idealThreadCount();
 		}
 		QThreadPool::globalInstance()->setMaxThreadCount(maxThreadCount);
 		QtConcurrent::blockingMap(cellsDescs, cloudMeshDistCellFunc_MT);
+#else
```
Comment (dgirardeau) [inline src/DistanceComputationTools.cpp:1798]: #elif defined(CC_CORE_LIB_USES_TBB)

**Review by dgirardeau — CHANGES_REQUESTED** [review]

**Review by andreadotti — COMMENTED** [review]
Addressed all comments

**Review by dgirardeau — APPROVED** [review]
Thanks for the updates! I generally prefer to be explicit, as it makes things more obvious when people are making local changes, and it's easier to detect inconsistencies.

**Comment by andreadotti** [comment]
Thank you @dgirardeau , any hope to get this merged and a release made out of this?

**Comment by dgirardeau** [comment]
Oh you could have merged, couldn't you?

**Comment by dgirardeau** [comment]
For the release, I'll simply update the submodule pointer in the CloudCompare repo (not sure how you generate CC on your side)

**Comment by dgirardeau** [comment]
Okay that's done on CC's side

**Comment by andreadotti** [comment]
Thank you very much for the merge. 
Actually I am considering using CCCoreLib library alone in an application, so I was referring to the possibility to create a release of this package only, I do not want to interfere with your CC workflow.
However if you do not want, it's not a big problem, I will refer to a specific commit in my building script

**Comment by dgirardeau** [comment]
Ah, we don't release CCCoreLib as a single component.

**Pull request merged**